ray-operator/controllers/ray/rayservice_controller_test.go

```diff
@@ -467,104 +360,149 @@ var _ = Context("RayService env tests", func() {
+			It("Should reflect the changes in the active cluster's WorkerGroupSpecs", func() {
 				Eventually(
 					getActiveRayClusterWorkerGroupSpecsFunc(ctx, rayService),
-					time.Second*15, time.Millisecond*500).Should(HaveLen(2))
+					time.Second*10, time.Millisecond*500).Should(HaveLen(2))
 			})
 
-			It("doesn't create a new pending cluster", func() {
+			It("Should not create a new pending cluster", func() {
 				Expect(rayService.Status.PendingServiceStatus.RayClusterName).To(BeEmpty())
 			})
 
-			It("doesn't switch to a new active cluster", func() {
-				Expect(rayService.Status.ActiveServiceStatus.RayClusterName).To(Equal(rayCluster.Name))
+			It("Should not switch to a new active cluster", func() {
+				clusterName := rayCluster.Name
+				Consistently(
+					getRayClusterNameFunc(ctx, rayService),
+					time.Second*3, time.Millisecond*500).Should(Equal(clusterName), "Active RayCluster: %v", rayService.Status.ActiveServiceStatus.RayClusterName)
 			})
 		})
 
-		When("during the zero-downtime upgrade", func() {
+		When("During the zero-downtime upgrade process", func() {
 			var pendingClusterName string
 
 			BeforeEach(OncePerOrdered, func() {
-				rayService.Spec.RayClusterSpec.RayVersion += "-next"
-				err := k8sClient.Update(ctx, rayService)
-				Expect(err).NotTo(HaveOccurred(), "failed to update RayService resource")
+				mockRayVersion := "mock-ray-version-1"
+				err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+					Eventually(
+						getResourceFunc(ctx, client.ObjectKey{Name: rayService.Name, Namespace: namespace}, rayService),
+						time.Second*3, time.Millisecond*500).Should(BeNil(), "RayService: %v", rayService.Name)
+					Expect(rayService.Spec.RayClusterSpec.RayVersion).NotTo(Equal(mockRayVersion))
+					rayService.Spec.RayClusterSpec.RayVersion = mockRayVersion
+					return k8sClient.Update(ctx, rayService)
+				})
+				Expect(err).NotTo(HaveOccurred(), "failed to update test RayService resource")
 
 				Eventually(
 					getPreparingRayClusterNameFunc(ctx, rayService),
 					time.Second*15, time.Millisecond*500).Should(Not(BeEmpty()), "Pending RayCluster name: %v", rayService.Status.PendingServiceStatus.RayClusterName)
 				pendingClusterName = rayService.Status.PendingServiceStatus.RayClusterName
+
+				// Wait until the pending RayCluster is created. If we update the RayService again before the cluster is created,
+				// the RayCluster will be constructed based on the new RayService spec and zero-downtime upgrade will not be triggered.
+				By(fmt.Sprintf("Waiting until the pending RayCluster %v is created", pendingClusterName))
+				pendingCluster := &rayv1.RayCluster{}
+				Eventually(
+					getResourceFunc(ctx, client.ObjectKey{Name: pendingClusterName, Namespace: namespace}, pendingCluster),
+					time.Second*15, time.Millisecond*500).Should(BeNil(), "Pending RayCluster: %v", pendingCluster.Name)
 			})
 
-			When("updating the RayVersion again", Ordered, func() {
```

Comment:
This test initially verifies an incorrect behavior:

(1) In `BeforeEach`, `RayVersion` is updated, which sets the pending cluster name in the RayService status.
(2) Then, in `"updating the RayVersion again"`, `RayVersion` is updated once more.

However, the RayCluster reconciler has not yet run, so the RayCluster is not created. Consequently, when the RayCluster finally reconciles, it uses the name generated in step (1) and derives its spec from the RayService spec set in step (2). As a result, no new pending cluster is created.

To make sure the old pending RayCluster use the spec from (1), `BeforeEach` will make sure the pending cluster is created.